libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -582,12 +612,67 @@ void AC_PosControl::input_vel_accel_xy(Vector2f& vel, const Vector2f& accel, boo
 ///     The parameter limit_output specifies if the velocity and acceleration limits are applied to the sum of commanded and correction values or just correction.
 void AC_PosControl::input_pos_vel_accel_xy(Vector2p& pos, Vector2f& vel, const Vector2f& accel, bool limit_output)
 {
+    // remove offsets including terrain offset for flat earth assumption
+    _pos_target.xy() -= _pos_offset.xy();
+    _vel_desired.xy() -= _vel_offset.xy();
+    _accel_desired.xy() -= _accel_offset.xy();
+
     update_pos_vel_accel_xy(_pos_target.xy(), _vel_desired.xy(), _accel_desired.xy(), _dt, _limit_vector.xy(), _p_pos_xy.get_error(), _pid_vel_xy.get_error());
 
     shape_pos_vel_accel_xy(pos, vel, accel, _pos_target.xy(), _vel_desired.xy(), _accel_desired.xy(),
                            _vel_max_xy_cms, _accel_max_xy_cmss, _jerk_max_xy_cmsss, _dt, limit_output);
 
     update_pos_vel_accel_xy(pos, vel, accel, _dt, Vector2f(), Vector2f(), Vector2f());
+
+    // update the position, velocity and acceleration offsets
+    update_xy_offsets();
+
+    // add back offsets including terrain offset
+    _pos_target.xy() += _pos_offset.xy();
+    _vel_desired.xy() += _vel_offset.xy();
+    _accel_desired.xy() += _accel_offset.xy();
+}
+
+/// update the horizontal position and velocity offsets
+/// this moves the offsets (e.g _pos_offset, _vel_offset, _accel_offset) towards the targets (e.g. _pos_offset_target or _vel_offset_target)
+void AC_PosControl::update_xy_offsets()
+{
+    switch (_xy_offset_type) {
+
+    case XYOffsetType::NONE:
+        break;
+
+    case XYOffsetType::POSITION: {
```

Comment:
Do we need a check to zero the position offset if we don't get an update for an extended time?

Reply:
I'm not sure if we need that or not.  For guided mode control (which I know is a little different) we don't timeout the position, we only time out velocity commands.

Reply:
If we timeout the velocity offset target (and zero it), we should also zero the accel offset target at the same time right?

Reply:
I've added a single timeout of 3 seconds for pos, vel and accel